### PR TITLE
fix: teardown and restart recording on session id change

### PR DESCRIPTION
**Pull request by pauldambra**
sometimes we see the artificial full snapshot on session idle timeout end up in the previous session not the new session - no bueno

in an attempt to not keep adding conditionals here i want to explore if tearing down rrweb and restarting recording is more stable

---

locally this seems way more reliable, but I want to get https://github.com/PostHog/posthog-js/pull/1416 in so I can retest without that noise

### CODE REVIEW

File: src/extensions/replay/sessionrecording.ts [hunk 1]
```diff
@@ -579,12 +590,11 @@ export class SessionRecording {
...
 
-        if (
-            returningFromIdle ||
-            ([FULL_SNAPSHOT_EVENT_TYPE, META_EVENT_TYPE].indexOf(event.type) === -1 &&
-                (windowIdChanged || sessionIdChanged || isUndefined(this._fullSnapshotTimer)))
-        ) {
-            this._tryTakeFullSnapshot()
+        if (sessionIdChanged || windowIdChanged) {
+            this.stopRecording()
+            this.startIfEnabledOrStop()
```
Comment (daibhin) [inline src/extensions/replay/sessionrecording.ts:595]: Haven't looked into it but wondering if we need to consider anything here with regards to ingestion controls (sampling, feature flags, minimum duration) that might affect the new recording not starting

**Review by daibhin — COMMENTED** [review]
Looks like a simple enough change, happy to approve once you're ready for a full review

File: src/extensions/replay/sessionrecording.ts (hunk 1 above)
Comment (pauldambra) [reply src/extensions/replay/sessionrecording.ts:595]: sampling and feature flags should just work...

min duration is interesting... i don't know if we'd restart the counter - we should... 👀

File: src/__tests__/extensions/replay/sessionrecording.test.ts [hunk 2]
```diff
@@ -426,26 +427,6 @@ describe('SessionRecording', () => {
             })
         })
 
-        it('when the first event is an incremental it does take a manual full snapshot', () => {
```
Comment (pauldambra) [inline src/__tests__/extensions/replay/sessionrecording.test.ts:429]: not manually fangling this any more so we don't need to test it happens

File: src/__tests__/extensions/replay/sessionrecording.test.ts [hunk 3]
```diff
@@ -764,9 +744,8 @@ describe('SessionRecording', () => {
             expect(posthog.capture).toHaveBeenCalledWith(
                 '$snapshot',
                 {
-                    $snapshot_bytes: 80,
+                    $snapshot_bytes: 60,
                     $snapshot_data: [
-                        createFullSnapshot(),
```
Comment (pauldambra) [inline src/__tests__/extensions/replay/sessionrecording.test.ts:769]: throughout these tests we're no longer manually adding full snapshots so we don't have to assert their presence

**Review by daibhin — APPROVED** [review]

File: src/extensions/replay/sessionrecording.ts (hunk 1 above)
Comment (pauldambra) [reply src/extensions/replay/sessionrecording.ts:597]: yep... session start is read from storage, most recent snapshot data from the buffer, so we're safe with min duration  too 👍

**Pull request merged**